pricegraph/src/orderbook.rs

```diff
@@ -146,53 +146,93 @@ impl Orderbook {
     /// transitive order, reducing them both. Returns the computed overlapping
     /// transitive orderbook for that market.
     ///
-    /// Note that there may exist additional transitive orders that may overlap
-    /// with orders in the resulting transitive orderbook. However, all negative
-    /// cycles will be removed from the orderbook graph.
+    /// Note that there may exist additional transitive orders that overlap with
+    /// orders in the resulting transitive orderbook. However, all negative
+    /// cycles over the specified market will be removed from the orderbook
+    /// graph. That is, the subgraph(s) containing the `quote` and `base` tokens
+    /// are guaranteed to be fully reduced.
     pub fn reduce_overlapping_transitive_orderbook(
         &mut self,
         market: Market,
     ) -> TransitiveOrderbook {
+        let mut overlap = TransitiveOrderbook::default();
+        while let Some(Ring { ask, bid }) = self.fill_market_ring_trade(market) {
+            overlap.asks.push(ask.as_transitive_order());
+            overlap.bids.push(bid.as_transitive_order());
+        }
+
+        // NOTE: In the case where the market `quote` and `base` token are in
+        // different disconnected subgraphs, it is possible that the `base`
+        // token's subgraph contains negative cycles. In order to ensure that
+        // the `base` token subgraph is also reduced, fill any remaining
+        // negative cycles in the inverse market. However, there should be no
+        // ring trades over the inverse market (if there were, then the `quote`
+        // and `base` token would be part of the same subgraph), so assert it.
+        let inverse_ring = self.fill_market_ring_trade(Market {
+            base: market.quote,
+            quote: market.base,
+        });
+        debug_assert_eq!(inverse_ring, None);
+
+        overlap
+    }
+
+    /// Fills a ring trade over the specified market, and returns the flow
+    /// corresponding to both ask and bid segments of the ring. Returns `None`
+    /// if there are no overlapping ring trades over the specified market.
+    ///
+    /// Note that if this method returns `None`, then the orderbook is
+    /// **partially** reduced. Specifically, the subgraph containing the market
+    /// `quote` token is fully reduced, however, other not connected subgraphs,
+    /// specifically the market `base`'s subgraph in the case where the `quote`
+    /// and `base` token are not part of the same subgraph, may still contain
+    /// negative cycles.
+    pub fn fill_market_ring_trade(&mut self, market: Market) -> Option<Ring> {
         if !self.is_token_pair_valid(market.bid_pair()) {
             return Default::default();
```

Comment:
feels more natural with the changed signature
```suggestion
            return None;
```